File: July 2024/22-07-2024/Largest BST.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

class Node
{
    constructor(x){
        this.key=x;
        this.left=null;
        this.right=null;
    }
}

function buildTree(str)
{
	// Corner Case
	if (str.length === 0 || str[0] == 'N')
		return null;

	// Creating vector of strings from input
	// string after spliting by space
	let ip = new Array();
    
    let ip_str = str.split(" ");
    
	for (let i=0;i<ip_str.length;i++)
		ip.push(ip_str[i]);

	// Create the root of the tree
    let root = new Node(parseInt(ip[0]));

	// Push the root to the queue
	let queue = new Array();
	queue.push(root);

	// Starting from the second element
	let i = 1;
	while (queue.length!==0 && i < ip.length) {

		// Get and remove the front of the queue
		let currNode = queue[0];
		queue.shift();

		// Get the current node's value from the string
		let currVal = ip[i];

		// If the left child is not null
		if (currVal != "N") {

			// Create the left child for the current node
			currNode.left = new Node(parseInt(currVal));

			// Push it to the queue
			queue.push(currNode.left);
		}

		// For the right child
		i++;
		if (i >= ip.length)
			break;
		currVal = ip[i];

		// If the right child is not null
		if (currVal != "N") {

			// Create the right child for the current node
			currNode.right = new Node(parseInt(currVal));

			// Push it to the queue
			queue.push(currNode.right);
		}
		i++;
	}

	return root;
}



function main() {
    let t = parseInt(readLine());
    for(let i=0;i<t;i++)
    {
        let s_tree = readLine().trim();
        let root = buildTree(s_tree);
        let obj = new Solution();
       
        let res = obj.largestBst(root);
        console.log(res);
    }
}

// } Driver Code Ends


//User function Template for javascript

/*
class Node
{
    constructor(x){
        this.key=x;
        this.left=null;
        this.right=null;
    }
}
*/

/**
 * @param {Node} root
 * @return {number}
*/


//Your Code Start
class BST {
    constructor(isBST, size, min, max) {
        this.isBST = isBST;
        this.size = size;
        this.min = min;
        this.max = max;
    }
}

class Solution {
    static subHelp(root, maxSize) {
        if (root === null) {
            return new BST(true, 0, Number.MAX_SAFE_INTEGER, Number.MIN_SAFE_INTEGER);
        }

        let left = Solution.subHelp(root.left, maxSize);
        let right = Solution.subHelp(root.right, maxSize);

        let curr = new BST(false, 0, 0, 0);

        if (left.isBST && right.isBST && root.key > left.max && root.key < right.min) {
            curr.isBST = true;
            curr.size = left.size + 1 + right.size;
            curr.min = Math.min(root.key, left.min);
            curr.max = Math.max(root.key, right.max);
            maxSize[0] = Math.max(maxSize[0], curr.size);
        } else {
            curr.size = Math.max(left.size, right.size);
        }

        return curr;
    }
    
   largestBst(root) {
        let maxSize = [0];
        Solution.subHelp(root, maxSize);
        return maxSize[0];
    }
}
//Your Code End

File: July 2024/22-07-2024/Largest BST.py
#User function Template for python3


#Your Code Start
class BST:
    def __init__(self, is_bst, size, min_val, max_val):
        self.is_bst = is_bst
        self.size = size
        self.min = min_val
        self.max = max_val

class Solution:
    @staticmethod
    def largestBst(root):
        max_size = [0]
        Solution.sub_help(root, max_size)
        return max_size[0]

    @staticmethod
    def sub_help(root, max_size):
        if root is None:
            return BST(True, 0, float('inf'), float('-inf'))

        left = Solution.sub_help(root.left, max_size)
        right = Solution.sub_help(root.right, max_size)

        curr = BST(False, 0, 0, 0)

        if left.is_bst and right.is_bst and root.data > left.max and root.data < right.min:
            curr.is_bst = True
            curr.size = left.size + 1 + right.size
            curr.min = min(root.data, left.min)
            curr.max = max(root.data, right.max)
            max_size[0] = max(max_size[0], curr.size)
        else:
            curr.size = max(left.size, right.size)

        return curr
#Your Code End



#{ 
 # Driver Code Starts
from collections import deque


# Tree Node
class Node:

    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


# Function to Build Tree
def buildTree(s):
    # Corner Case
    if len(s) == 0 or s[0] == "N":
        return None

    # Creating list of strings from input string after splitting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size += 1

    # Starting from the second element
    i = 1
    while size > 0 and i < len(ip):
        # Get and remove the front of the queue
        currNode = q.popleft()
        size -= 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if currVal != "N":
            # Create the left child for the current node
            currNode.left = Node(int(currVal))
            # Push it to the queue
            q.append(currNode.left)
            size += 1

        # Move to the next element
        i += 1
        if i >= len(ip):
            break

        currVal = ip[i]

        # If the right child is not null
        if currVal != "N":
            # Create the right child for the current node
            currNode.right = Node(int(currVal))
            # Push it to the queue
            q.append(currNode.right)
            size += 1

        # Move to the next element
        i += 1

    return root


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        s = input()
        root = buildTree(s)
        ob = Solution()
        result = ob.largestBst(root)
        print(result)

# } Driver Code Ends
